File: src/python/dbl.twins.sdk.core/twinclient.py
import json
import asyncio
from typing import List, Callable

# Models
class RoomPatch:
    def __init__(self, value: int, path: str, op: str):
        self.value = value
        self.path = path
        self.op = op

class RoomModel:
    def __init__(self, model_id: str, patch: List[RoomPatch]):
        self.model_id = model_id
        self.patch = patch

# Event arguments
class ConnectionEventArgs:
    def __init__(self, is_connected: bool, message: str):
        self.is_connected = is_connected
        self.message = message

class TelemetryEventArgs:
    def __init__(self, source: str, telemetry_data: str):
        self.source = source
        self.telemetry_data = telemetry_data

# Main Client
class TwinClient:
    def __init__(self, event_hub_connection_string: str):
        self._connection_string = event_hub_connection_string
        parts = self._connection_string.split(';')
        self._event_hub_name = parts[0]
        
        # Events represented as lists of callable functions
        self.telemetry_update_callbacks: List[Callable[[TelemetryEventArgs], None]] = []
        self.connected_callbacks: List[Callable[[ConnectionEventArgs], None]] = []

    async def connect_hub(self):
        # ... (Your Azure EventHubs Python SDK code here)

        # This is just a placeholder to show how you might handle events
        data_received = "{...}"  # Placeholder for the data received from Azure EventHubs
        self._on_telemetry_update("SomeSubject", data_received)

    def _on_connected(self, is_connected: bool, msg: str):
        event_args = ConnectionEventArgs(is_connected, msg)
        for callback in self.connected_callbacks:
            callback(event_args)

    def _on_telemetry_update(self, subject: str, data: str):
        room_model = json.loads(data, cls=RoomModel)
        print(f"Model ID: {room_model.model_id}")
        
        for patch in room_model.patch:
            print(f"Operation: {patch.op}")
            print(f"Path: {patch.path}")
            print(f"Value: {patch.value}")
            
            event_args = TelemetryEventArgs(subject, str(patch.value))
            for callback in self.telemetry_update_callbacks:
                callback(event_args)

# Usage
client = TwinClient("YourConnectionStringHere")

def telemetry_updated(args: TelemetryEventArgs):
    print(f"Telemetry Updated: {args.telemetry_data}")

client.telemetry_update_callbacks.append(telemetry_updated)
asyncio.run(client.connect_hub())

File: src/csharp/dbl.twins.sdk.core/TwinClient.cs
﻿using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using Azure.Messaging.EventHubs;
using Azure.Messaging.EventHubs.Consumer;
using Azure.Messaging.EventHubs.Producer;
using Newtonsoft.Json;

namespace dbl.twins.sdk.core
{

    //Room Updates
    public class RoomPatch
    {
        public int Value { get; set; }
        public string Path { get; set; }
        public string Op { get; set; }
    }

    public class RoomModel
    {
        public string ModelId { get; set; }
        public RoomPatch[] Patch { get; set; }
    }



    public class ConnectionEventArgs : EventArgs
    {
        public bool isConnected;
        public string message;

        public ConnectionEventArgs(bool isConnected, string message)
        {
            this.isConnected = isConnected;
            this.message = message;
        }
    }

    public class TelemetryEventArgs : EventArgs
    {
        public string source;
        public string telemetryData;

        public TelemetryEventArgs(string source, string telemetryData)
        {
            this.source=source;
            this.telemetryData=telemetryData;
        }
    }


    /// <summary>
    /// Client that connects to Event Hub to recieve telmetry events
    /// </summary>
    public class TwinClient
    {
        private string _connectionString;
        private string _eventHubName;
        public event EventHandler<TelemetryEventArgs> TelemetryUpdate;
        public event EventHandler<ConnectionEventArgs> Connected;

        public TwinClient(string eventHubConnectionString)
        {
            _connectionString = eventHubConnectionString;
            var parts = _connectionString.Split(new char[] { ';' });
            _eventHubName = parts[0];

        }

        public async Task ConnectHub()
        {
            var consumerGroup = EventHubConsumerClient.DefaultConsumerGroupName;
            var consumer = new EventHubConsumerClient(consumerGroup, _connectionString);

            try
            {
                OnConnected(true, "Connected");
                using CancellationTokenSource cancellationSource = new CancellationTokenSource();
                //cancellationSource.CancelAfter(TimeSpan.FromSeconds(45));

                int eventsRead = 0;
                int maximumEvents = 30000;
                int count = 0;

                await foreach (PartitionEvent partitionEvent in consumer.ReadEventsAsync(cancellationSource.Token))
                {

                    count++;
                    string readFromPartition = partitionEvent.Partition.PartitionId;
                    byte[] eventBodyBytes = partitionEvent.Data.EventBody.ToArray();
                    string eventBodyString = Encoding.UTF8.GetString(eventBodyBytes);

                    string subject = "";
                    if (partitionEvent.Data.Properties.ContainsKey("cloudEvents:subject"))
                    {
                        subject = partitionEvent.Data.Properties["cloudEvents:subject"].ToString();
                        OnTelemetryUpdate(subject, eventBodyString);
                    }

                    Console.WriteLine($"{count} - Read event {subject} : {eventBodyString}");


                    eventsRead++;

                    if (eventsRead >= maximumEvents)
                    {
                        OnConnected(false, "Maximum Events reached");
                        break;
                    }
                }
            }
            catch (TaskCanceledException)
            {
                OnConnected(false, "Cancelled");
                // This is expected if the cancellation token is
                // signaled.
            }
            catch (System.Exception ex)
            {
                OnConnected(false, ex.ToString());
            }
            finally
            {
                OnConnected(false, "Connection Closed");
                await consumer.CloseAsync();
            }
        }

        private void OnConnected(bool isConnected, string msg)
        {
            EventHandler<ConnectionEventArgs> handler = Connected;
            if (null != handler)
            {
                Connected(this, new ConnectionEventArgs(isConnected, msg));
            }
        }

        //Raise Telemetry Event
        private void OnTelemetryUpdate(string subject, string data)
        {
            EventHandler<TelemetryEventArgs> handler = TelemetryUpdate;
            if (null != handler)
            {
                var roomModel = JsonConvert.DeserializeObject<RoomModel>(data);

                Console.WriteLine("Model ID: " + roomModel.ModelId);

                foreach (var patch in roomModel.Patch)
                {
                    Console.WriteLine("Operation: " + patch.Op);
                    Console.WriteLine("Path: " + patch.Path);
                    Console.WriteLine("Value: " + patch.Value);

                    handler(this, new TelemetryEventArgs(subject, patch.Value.ToString()));
                }
            }
        }
    }
}
